File: JAVA/09_TextProcessing/Lecture42/Trie.java
// Trie - data Structure

package Lecture42;

import java.util.ArrayList;
import java.util.HashMap;

public class Trie {
	
	static class Node{
		char data;
		boolean eow;
		HashMap<Character, Node> cmap = new HashMap<>();
		Node(char data, boolean eow){
			this.data = data;
			this.eow = eow;
		}
	}
	
	static void addWord(Node root, String word) {
		if(word.length() == 0) {
//			System.out.println("Inside");
			root.eow = true;
			return;
		}
		char ch = word.charAt(0);
		String row = word.substring(1);
		if(root.cmap.containsKey(ch) == true) {
			Node child = root.cmap.get(ch);
			addWord(child, row);
		} else {
			Node child = new Node(ch, false);
			root.cmap.put(ch, child);
			addWord(child, row);
		}
	}
	
	static void removeWord(Node root, String word) {
		if(word.length() == 0) {
			root.eow = false;
			return;
		}
		char ch = word.charAt(0);
		String row = word.substring(1);
		Node child = root.cmap.get(ch);
		removeWord(child, row);
		
		if(child.eow == false && child.cmap.size() == 0) {
			root.cmap.remove(ch);
		}
		
		
	}
	
	static void displayWord(Node root, String asf) {
		if(root.eow == true) {
			System.out.print(asf + ", ");
		}
		ArrayList<Character> ans = new ArrayList<>(root.cmap.keySet());
		for(int i = 0; i < ans.size(); i++) {
			Node child = root.cmap.get(ans.get(i));
			displayWord(child, asf + ans.get(i));
		}
	}
	
	static void displayTrie(Node root) {
		System.out.print("[" + root.data + " " + root.eow  + "] -> ");
		ArrayList<Character> child = new ArrayList<>(root.cmap.keySet());
		for(int i = 0; i < child.size(); i++) {
			System.out.print(child.get(i) + ", ");
		}
		System.out.println(".");
		for(int i = 0; i < child.size(); i++) {
			Node cd = root.cmap.get(child.get(i));
			displayTrie(cd);
		}
	}
	
	static boolean searchWord(Node root, String word) {
		
		return false;
	}
	
	public static void main(String[] args) {
		Node root = new Node('$', false);
		addWord(root, "an");
		addWord(root, "and");
		addWord(root, "ant");
		addWord(root, "are");
		addWord(root, "as");
		addWord(root, "ask");
		addWord(root, "see");
		addWord(root, "sea");
		addWord(root, "seen");
		System.out.print("[");
		displayWord(root, "");
		System.out.println("]");
		displayTrie(root);
		System.out.println("~~~~~~~~~~~~~~~~~~~");
		removeWord(root, "as");
		displayTrie(root);
		System.out.println("~~~~~~~~~~~~~~~~~~~");
		removeWord(root, "ask");
		displayTrie(root);
		System.out.println("~~~~~~~~~~~~~~~~~~~");
	}
}

File: C++/DS2/Lec42/trie.cpp
#include<iostream>
#include<unordered_map>
#include<cstring>
#include<vector>

using namespace std;

class Node
{
    public:
        char data;
        bool eow;   // end of word show the existence
        unordered_map<char, Node*> cmap;

        Node(char data, bool eow)
        { 
            this->data = data;
            this->eow = eow;
        }
};

bool searchWord(Node* root, string word)
{
    if(word.size() == 0)
    {
        return root->eow;
    }
    char ch = word[0];
    string row = word.substr(1);
    if(root->cmap.count(ch) == 0){
        return false;
    }
    else 
    {
        Node* child = root->cmap[ch];
        return searchWord(child, row);
    }
}

void addWord(Node* root, string word)
{
    if(word.size() == 0)
    {
        root->eow = true;
        return;
    }
    char cw = word[0];
    string row = word.substr(1);
    if(root->cmap.count(cw) == 0)
    {
        Node* child = new Node(cw, false);
        root->cmap[cw] = child;
        addWord(child, row);
    }
    else
    {
        addWord(root->cmap[cw], row);
    }
    
}

void removeWord(Node* root, string word)
{
    if(word.size() == 0)
    {
        root->eow = false;
        return;
    }
    char ch = word[0];
    string row = word.substr(1);
    Node* child = root->cmap[ch];
    removeWord(child, row);

    if(child->eow == false && child->cmap.size() == 0)
    {
        root->cmap.erase(ch);
        delete child;
    }
}

void displayWords(Node* root, string wsf)
{
    if(root->eow == true)
    {
        cout << wsf << ", ";
    }
    for(auto itr = root->cmap.begin(); itr != root->cmap.end(); itr++)
    {
        displayWords(itr->second, wsf + itr->first);
    }
}

void displayTrie(Node* root)
{
    cout << "[" << root->data  << " " << root->eow << "] - ";
    for(auto itr = root->cmap.begin(); itr != root->cmap.end(); itr++)
    {
        cout << itr->first << ", ";
    }
    cout << "." << endl;
    for(auto itr = root->cmap.begin(); itr != root->cmap.end(); itr++)
    {
        displayTrie(itr->second);
    }
}


int main(int argv, char** argc)
{
    Node* root = new Node('$', false);
    addWord(root, "are");
    addWord(root, "art");
    addWord(root, "as");
    addWord(root, "ask");
    addWord(root, "ant");
    addWord(root, "an");
    addWord(root, "and");
    addWord(root, "seen");
    addWord(root, "sea");
    addWord(root, "see");
    // displayWords(root, "");
    // cout << endl;
    // removeWord(root, "as");
    // displayTrie(root);
    // cout << "~~~~~~~~~~~~~~~~~" << endl;
    // removeWord(root, "ask");
    // displayTrie(root);
    cout << searchWord(root, "seen") << endl;

}
